src/cdf.rs

```diff
@@ -5075,66 +5085,40 @@ unsafe fn get_qcat_idx(q: c_int) -> c_int {
-    if !((*src).r#ref).is_null() {
-        memcpy(
-            dst as *mut c_void,
-            (*src).data.cdf as *const c_void,
-            ::core::mem::size_of::<CdfContext>(),
-        );
-    } else {
-        (*dst).m = av1_default_cdf.clone();
-        (*dst).kfym = default_kf_y_mode_cdf;
-        (*dst).coef = av1_default_coef_cdf[(*src).data.qcat as usize].clone();
-        (*dst).mv.joint = default_mv_joint_cdf;
-        (*dst).dmv.joint = default_mv_joint_cdf;
-        (*dst).dmv.comp[1] = default_mv_component_cdf.clone();
-        (*dst).dmv.comp[0] = (*dst).dmv.comp[1].clone();
-        (*dst).mv.comp[1] = (*dst).dmv.comp[0].clone();
-        (*dst).mv.comp[0] = (*dst).mv.comp[1].clone();
-    };
+pub unsafe fn rav1d_cdf_thread_copy(dst: &mut CdfContext, src: &CdfThreadContext) {
+    match src {
+        CdfThreadContext::Cdf(src) => {
+            *dst = src.cdf.try_read().unwrap().clone();
+        }
+        CdfThreadContext::QCat(i) => {
+            dst.m = av1_default_cdf.clone();
+            dst.kfym = default_kf_y_mode_cdf;
+            dst.coef = av1_default_coef_cdf[*i as usize].clone();
+            dst.mv.joint = default_mv_joint_cdf;
+            dst.dmv.joint = default_mv_joint_cdf;
+            dst.dmv.comp[1] = default_mv_component_cdf.clone();
+            dst.dmv.comp[0] = dst.dmv.comp[1].clone();
+            dst.mv.comp[1] = dst.dmv.comp[0].clone();
+            dst.mv.comp[0] = dst.mv.comp[1].clone();
+        }
+    }
 }
 
 pub unsafe fn rav1d_cdf_thread_alloc(
-    c: &Rav1dContext,
-    cdf: *mut CdfThreadContext,
+    _c: &Rav1dContext,
     have_frame_mt: c_int,
-) -> Rav1dResult {
-    (*cdf).r#ref = rav1d_ref_create_using_pool(
-        c.cdf_pool,
-        (::core::mem::size_of::<CdfContext>()).wrapping_add(::core::mem::size_of::<AtomicU32>()),
-    );
-    if ((*cdf).r#ref).is_null() {
-        return Err(ENOMEM);
-    }
-    (*cdf).data.cdf = (*(*cdf).r#ref).data as *mut CdfContext;
-    if have_frame_mt != 0 {
-        (*cdf).progress = &mut *((*cdf).data.cdf).offset(1) as *mut CdfContext as *mut AtomicU32;
-        (*cdf).progress.write(AtomicU32::new(0));
-    }
-    Ok(())
-}
-
-pub unsafe fn rav1d_cdf_thread_ref(dst: *mut CdfThreadContext, src: *mut CdfThreadContext) {
-    *dst = (*src).clone();
-    if !((*src).r#ref).is_null() {
-        rav1d_ref_inc((*src).r#ref);
-    }
-}
-
-pub unsafe fn rav1d_cdf_thread_unref(cdf: &mut CdfThreadContext) {
-    *cdf = CdfThreadContext {
-        r#ref: cdf.r#ref,
-        data: CdfThreadContext_data {
-            cdf: ptr::null_mut(),
-            // cdf is larger than qcat, so this zeroes it
-        },
-        progress: ptr::null_mut(),
+) -> Rav1dResult<CdfThreadContext> {
+    let progress = if have_frame_mt != 0 {
+        Some(AtomicU32::new(0))
+    } else {
+        None
     };
-    rav1d_ref_dec(&mut cdf.r#ref);
+    Ok(CdfThreadContext::Cdf(Arc::new(CdfThreadContext_data {
+        cdf: mem::zeroed(),
```

Comment:
Why do we need to introduce a new `mem::zeroed()`, or is this split off an existing, larger `mem::zeroed()`?

Reply:
Arc needs to not be zero-initialized, so this is split from the parent's zeroed init.